src/coord/src/catalog/storage.rs

```diff
@@ -253,168 +255,154 @@ impl Connection {
-            )
-            .optional()?;
+    fn get_setting(&self, key: &str) -> Result<Option<String>, Error> {
+        Self::get_setting_stash(&self.stash, key)
+    }
+
+    fn set_setting(&self, key: &str, value: &str) -> Result<(), Error> {
+        Self::set_setting_stash(&self.stash, key, value)
+    }
 
-        let res = match current_setting {
+    fn get_setting_stash(stash: &impl Stash, key: &str) -> Result<Option<String>, Error> {
+        let settings = COLLECTION_SETTING.get(stash)?;
+        let v = stash.peek_key_one(
+            settings,
+            &SettingKey {
+                name: key.to_string(),
+            },
+        )?;
+        Ok(v.map(|v| v.value))
+    }
+
+    fn set_setting_stash<V: Into<String> + std::fmt::Display>(
+        stash: &impl Append,
+        key: &str,
+        value: V,
+    ) -> Result<(), Error> {
+        let key = SettingKey {
+            name: key.to_string(),
+        };
+        let value = SettingValue {
+            value: value.into(),
+        };
+        COLLECTION_SETTING
+            .upsert(stash, once((key, value)))
+            .map_err(|e| e.into())
+    }
+
+    /// Sets catalog's `cluster_id` setting on initialization or gets that value.
+    fn set_or_get_cluster_id(stash: &impl Append) -> Result<Uuid, Error> {
+        let current_setting = Self::get_setting_stash(stash, "cluster_id")?;
+        match current_setting {
             // Server init
             None => {
                 // Generate a new version 4 UUID. These are generated from random input.
                 let cluster_id = Uuid::new_v4();
-                tx.execute(
-                    "INSERT INTO settings VALUES ('cluster_id', ?);",
-                    params![SqlVal(cluster_id)],
-                )?;
+                Self::set_setting_stash(stash, "cluster_id", cluster_id.to_string())?;
                 Ok(cluster_id)
             }
             // Server reboot
-            Some(cs) => Ok(cs.0),
-        };
-        tx.commit()?;
-        res
+            Some(cs) => Ok(Uuid::parse_str(&cs)?),
+        }
     }
 
     pub fn get_catalog_content_version(&mut self) -> Result<String, Error> {
-        let tx = self.inner.transaction()?;
-        let current_setting: Option<String> = tx
-            .query_row(
-                "SELECT value FROM settings WHERE name = 'catalog_content_version';",
-                params![],
-                |row| row.get(0),
-            )
-            .optional()?;
-        let version = match current_setting {
-            Some(v) => match v.parse::<u32>() {
-                // Prior to v0.8.4 catalog content versions was stored as a u32
-                Ok(_) => "pre-v0.8.4".to_string(),
-                Err(_) => v,
-            },
-            None => "new".to_string(),
-        };
-        tx.commit()?;
-        Ok(version)
+        self.get_setting("catalog_content_version")
+            .map(|v| v.unwrap_or_else(|| "new".to_string()))
```

Comment:
I agree this is weird, and I also didn't understand what it was trying to do when I read it. I believe this is a thing that should be addressed in a followup PR because it's existing behavior.